B-Sockets-ServerTesting/sockets-servertesting.md

```diff
@@ -0,0 +1,285 @@
+    - "\r\n"
+
+**HTTP response**
+- First line: response status
+    - Success: "HTTP/1.0 200 OK\r\n"
+    - Failure: "HTTP/1.0 404 Not Found\r\n"
+- Followed by 0 or more response headers
+- Followed by an empty line
+- Followed by the content of the response
+    - Ex: image file or HTML file
+
+We can see examples of these two HTTP requests with `netcat`. Since both
+`netcat` and web servers/clients connect over sockets, we can pretend to be a
+web client or server and recieve the text that the opposing computer sends.
+
+Let's first act as a web server. We first tell `netcat` to open a server
+connection with `nc -l 10000`, and then in a web browser navigate to the URL
+that specifies this server. You can use the domain name `localhost` to specify
+the local computer rather than connecting to a remote computer over the
+internet. In Chrome, I'll navigate to the URL
+`http://localhost:10000/index.html`. `netcat` outputs this:
+```
+$ nc -l 10000
+GET /index.html HTTP/1.1   # GET == method; /index.html == request URI; HTTP/1.0 == version
+Host: localhost:10000      # header
+Connection: keep-alive     # more headers...
+-removed for brevity-
+                           # blank newline to indicate end of headers/request
+```
+
+To act as a client, we can type our HTTP request manually into netcat rather
+than doing it through the web browser. Here, I'll try to send an HTTP request to
+the domain name `example.com` on port `80` (the default for HTTP web servers)
+for the URI `/index.html`. Note that I specify the `-C` with `netcat` so that
+newlines are `\r\n` rather than `\n` -- a requirement of the HTTP protocol. This
+flag may be different depending on your operating system.
+```
+$ nc -C example.com 80
+GET /index.html HTTP/1.0    # GET == method; /index.html == request URI; HTTP/1.1 == version
+                            # blank line to specify end of request
+HTTP/1.0 200 OK             # start of HTTP response. HTTP/1.0 == version; 200 OK == response status
+Accept-Ranges: bytes        # header
+Content-Type: text/html     # more headers...
+-removed for brevity-
+                            # blank newline to indicate end of headers and start of file contents
+<!doctype html>
+<html>
+<head>
+    <title>Example Domain</title>
+-removed for brevity-
+```
+
+Testing your multi-server
+---------------------------
+### Siege
+Siege is a command-line tool that allows you to benchmark your webserver using
+load testing. Given a few parameters, Siege tells you things like the number of
+successful transactions to your website, percent availability, and
+latency/throughput of your server, 
+
+To install siege, type the following command:
+
+`sudo apt install siege`
+
+To use siege with your webserver in HW3, run your server and test with the
+following command:
+
+`siege http://<hostname>:<port>/<url>`
+
+This will run for an infinite amount of time. When you CTRL-C out of the
+command, a list of statistics will be outputted on your terminal.
```

Comment:
```suggestion
This will run for an infinite amount of time. When you <kbd>Ctrl</kbd>-<kbd>C</kbd>
out of the command, a list of statistics will be outputted on your terminal.
```